### ISSUE TITLE
CA1801: False negatives due to argument validation

### ISSUE THREAD

**Issue by bordecal** [dotnet/roslyn-analyzers#3752]
The current implementation of CA1801 does not exclude common argument validation approaches from detected parameter "usage", resulting in a potentially high false negative rate for a code base that makes heavy use of argument validation.

The rule would be quite a bit more useful if it included an option for ignoring validation "usage", including both inline validation and validation via helper methods (_à la_ https://docs.microsoft.com/en-us/dotnet/api/microsoft.requires).  Ideally, this would allow for a more than just `ValidatedNotNullAttribute` for identifying a validation helper method since not all argument validation involves null checks (e.g.: https://docs.microsoft.com/en-us/dotnet/api/microsoft.requires.range).

#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

[CA1801](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1801) (ReviewUnusedParameters)

#### Repro steps

When a parameter's only usage inside a method is for argument validation (whether inline or via a validation method), a CA1801 violation is not flagged.  For example, a CA1801 violation will not be flagged for the `x` parameter in either `DoSomething1` or `DoSomething2` below.  (The same result occurs even if `nameof()` is not used to retrieve the parameter name.)

```csharp
public void DoSomething1(object x)
{
    if (x == null)
    {
        throw new ArgumentNullException(nameof(x));
    }
}

public void DoSomething2(object x)
{
    Argument.NotNull(x, nameof(x));
}
```

where the following are present for `DoSomething2`:

```csharp
[AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)]
public sealed class ValidatedNotNullAttribute : Attribute
{
}

public static class Argument
{
    public static T NotNull<T>([ValidatedNotNull] this T value, string parameterName)
    {
        return value ?? throw new ArgumentNullException(parameterName);
    }
}
```

#### Expected behavior
Validation should not be sufficient to consider a parameter "used" for the purposes of CA1801.

#### Actual behavior
Use of a parameter for either inline validation or validation via a helper method hides a potential CA1801 problem.

**Comment by Evangelink** [comment]
`CA1801` is deprecated and will be replaced by `IDE0060`, I have just tested your code and `IDE0060` doesn't have the false-positive so I am going to close your ticket.

Thank you for the report!
